### PR TITLE
Compliance multi-tenant changes

### PR DESCRIPTION
**Pull request by Josh-Tigera**
## For PR author

- [x] Tests for change.
~- [ ] If changing pkg/apis/, run `make gen-files`~
~- [ ] If changing versions, run `make gen-versions`~

## For PR reviewers

A note for code reviewers - all pull requests must have the following:

- [ ] Milestone set according to targeted release.
- [ ] Appropriate labels:
  - `kind/bug` if this is a bugfix.
  - `kind/enhancement` if this is a a new feature.
  - `enterprise` if this PR applies to Calico Enterprise only.

### CODE REVIEW

**Review by caseydavenport — CHANGES_REQUESTED** [review]
Looking good! A few relatively minor comments. Main one is re: how network policy is structured.

File: pkg/render/compliance.go [hunk 1]
```diff
@@ -77,14 +77,54 @@ const (
...
 
-var (
-	ComplianceServerEntityRule            = networkpolicy.CreateEntityRule(ComplianceNamespace, ComplianceServerName, complianceServerPort)
-	ComplianceServerSourceEntityRule      = networkpolicy.CreateSourceEntityRule(ComplianceNamespace, ComplianceServerName)
-	ComplianceBenchmarkerSourceEntityRule = networkpolicy.CreateSourceEntityRule(ComplianceNamespace, ComplianceBenchmarkerName)
-	ComplianceControllerSourceEntityRule  = networkpolicy.CreateSourceEntityRule(ComplianceNamespace, ComplianceControllerName)
-	ComplianceSnapshotterSourceEntityRule = networkpolicy.CreateSourceEntityRule(ComplianceNamespace, ComplianceSnapshotterName)
-	ComplianceReporterSourceEntityRule    = networkpolicy.CreateSourceEntityRule(ComplianceNamespace, ComplianceReporterName)
-)
+func createComplianceSourceEntityRule(deploymentName string, tenantNamespaces []string) v3.EntityRule {
```
[the remark this answers is not in the document]
Comment (Josh-Tigera) [reply pkg/render/compliance.go:80]: Ah ok I didn't know about this I'll make this reversion. My original intent with passing through all tenant namespaces was to make rules work for components that didn't support multi-tenancy yet but would still run in a multi-tenant cluster, but you mentioned guardian is managed cluster only and dex we don't need  to support multi-tenancy at all so I can just get rid of this pattern entirely.

File: pkg/controller/compliance/compliance_controller.go [hunk 2]
```diff
@@ -328,14 +354,19 @@ func (r *ReconcileCompliance) Reconcile(ctx context.Context, request reconcile.R
 		return reconcile.Result{}, err
 	}
 
-	certificateManager, err := certificatemanager.Create(r.client, network, r.clusterDomain, common.OperatorNamespace())
+	var opts []certificatemanager.Option
+	if tenant != nil {
+		opts = append(opts, certificatemanager.WithTenant(tenant))
```
Comment (caseydavenport) [inline pkg/controller/compliance/compliance_controller.go:359]: The `WithTenant` option actually handles a nil tenant behind the scenes, so no need for the `tenant != nil` check here.

Comment (caseydavenport) [reply pkg/controller/compliance/compliance_controller.go:359]: Would be nice to include a `WithLogger` option here to pass the reqLogger context logger.

File: pkg/render/manager.go [hunk 3]
```diff
@@ -165,11 +165,12 @@ type ManagerConfiguration struct {
...
 	BindingNamespaces []string
 
-	// Whether or not to run the rendered components in multi-tenant mode.
+	// Whether to run the rendered components in multi-tenant mode.
 	Tenant          *operatorv1.Tenant
 	ExternalElastic bool
 
-	Manager *operatorv1.Manager
+	Manager          *operatorv1.Manager
+	TenantNamespaces []string
```
[the remark this answers is not in the document]
Comment (caseydavenport) [reply pkg/render/manager.go:173]: @Josh-Tigera still need to remove this field ^

File: pkg/controller/manager/manager_controller.go [hunk 4]
```diff
@@ -640,6 +640,11 @@ func (r *ReconcileManager) Reconcile(ctx context.Context, request reconcile.Requ
 		namespaces = append(namespaces, render.ManagerNamespace)
 	}
 
+	tenantNamespaces, err := utils.TenantNamespaces(ctx, r.client)
```
Comment (caseydavenport) [inline pkg/controller/manager/manager_controller.go:643]: I don't think this is needed any more

File: pkg/controller/manager/manager_controller.go [hunk 5]
```diff
@@ -662,6 +667,7 @@ func (r *ReconcileManager) Reconcile(ctx context.Context, request reconcile.Requ
 		Namespace:               helper.InstallNamespace(),
 		TruthNamespace:          helper.TruthNamespace(),
 		Tenant:                  tenant,
+		TenantNamespaces:        tenantNamespaces,
```
Comment (caseydavenport) [inline pkg/controller/manager/manager_controller.go:670]: nor this

File: pkg/render/compliance.go [hunk 6]
```diff
@@ -277,12 +273,12 @@ var (
 	complianceReplicas int32 = 1
 )
 
-const complianceServerPort = 5443
+const ComplianceServerPort = 5443
```
Comment (caseydavenport) [inline pkg/render/compliance.go:276]: Does this need to be public now? I can't see where it's used outside of this package (but might have missed something)

File: pkg/render/logstorage/linseed/linseed.go [hunk 7]
```diff
@@ -119,6 +119,8 @@ type Config struct {
 	Tenant          *operatorv1.Tenant
 	ExternalElastic bool
 
+	TenantNamespaces []string
```
[the remark this answers is not in the document]
Comment (caseydavenport) [reply pkg/render/logstorage/linseed/linseed.go:122]: @Josh-Tigera same here - can remove this field from the struct now

**Review by caseydavenport — CHANGES_REQUESTED** [review]
A few more small things. Looks like you have a couple of unused TenantNamespaces fields left over from earlier commits as well.

**Comment by marvin-tigera** [comment]
Removing "merge-when-ready" label due to new commits

**Comment by marvin-tigera** [comment]
Removing "merge-when-ready" label due to new commits

File: pkg/controller/compliance/compliance_controller.go [hunk 8]
```diff
@@ -64,69 +66,58 @@ func Add(mgr manager.Manager, opts options.AddOptions) error {
...
+	go utils.WaitToAddTierWatch(networkpolicy.TigeraComponentTierName, complianceController, k8sClient, log, tierWatchReady)
+	go utils.WaitToAddNetworkPolicyWatches(complianceController, k8sClient, log, []types.NamespacedName{
+		{Name: render.ComplianceAccessPolicyName, Namespace: installNS},
+		{Name: render.ComplianceServerPolicyName, Namespace: installNS},
+		{Name: networkpolicy.TigeraComponentDefaultDenyPolicyName, Namespace: installNS},
+	})
 
 	// Watch for changes to primary resource Compliance
-	err = c.Watch(&source.Kind{Type: &operatorv1.Compliance{}}, &handler.EnqueueRequestForObject{})
+	err = complianceController.Watch(&source.Kind{Type: &operatorv1.Compliance{}}, &handler.EnqueueRequestForObject{})
```
Comment (caseydavenport) [inline pkg/controller/compliance/compliance_controller.go:102]: I think this needs to be `eventHandler`

File: pkg/render/guardian.go [hunk 9]
```diff
@@ -431,25 +431,25 @@ func guardianAllowTigeraPolicy(cfg *GuardianConfiguration) (*v3.NetworkPolicy, e
 		{
 			Action:      v3.Allow,
 			Protocol:    &networkpolicy.TCPProtocol,
-			Source:      ComplianceBenchmarkerSourceEntityRule,
+			Source:      networkpolicy.CreateSourceEntityRule(ComplianceNamespace, ComplianceBenchmarkerName),
```
Comment (caseydavenport) [inline pkg/render/guardian.go:434]: I think for all of these single-tenant packages, you just want to use the `DefaultHelper` - https://github.com/Josh-Tigera/operator/blob/691931ed32299b33efcbd98c8c69d8b49f1ba760/pkg/render/common/networkpolicy/networkpolicy.go#L205

This way we still follow the same code path for single / multi-tenant code.

**Pull request merged**